### DISCUSSION TITLE
Can tidy V-Green Variable Speed Motor

### DISCUSSION THREAD

**Discussion by zombielinux** [tagyoureit/nodejs-poolController#686] [RS485]
I’ve just got a v-Green variable speed motor to replace the single speed on my Hayward super pump. 

It’s got an RS-485 port on it and advertises wide compatibility with other automation systems. 

Is there already support for these devices? If not, what can I do to help add support for them?

**Comment by tagyoureit** [comment]
Am I correct in assuming it is not hooked up to any automation system currently?  Easiest way is to snoop on the packets.  Actually, that's the second easiest way.  The first easiest is to see the protocol documented somewhere.  :-)  

If it has "wide compatibility" you might want to just try hooking up to to the RS-485 bus and installing it in Nixie (setup page -> pumps -> add a pump) like Pentair VS or Hayward Eco/Tristar VS and see if it responds accordingly.

**Reply by zombielinux** [reply]
Sorry this took a minute. 

I finally got it all connected up. 

It doesn't respond at the default address (0x15) and any of the options. 

Attached is their protocol: https://www.troublefreepool.com/attachments/gen3-epc-modbus-communication-protocol-_rev4-17-pdf.358807/

https://www.troublefreepool.com/threads/century-regal-vgreen-motor-automation.238733/ seems to have a lot of information on it.

**Reply by tagyoureit** [reply]
That thread is very interesting... we can already "talk" the basic Aqualink protocol but so far only to read the OCP not control the pumps standalone.  Also, Modbus is a different protocol over the RS-485 bus compared to Aqualink or Pentair.  It would be interesting to try to set the noted configuration protocol to "Pentair" and see if we can talk to the pump.  Is that something you could try?

**Reply by zombielinux** [reply]
I didn't realize I needed a standalone controller to interface with pumps. I've been happily driving my SWG with a direct RS-485 connection this whole time. 

There are three different Pentair options in Settings -> Pumps -> Pump Type. 

I've currently got it set to "Intelliflow VS"

**Reply by tagyoureit** [reply]
You need some type of controller.  It can be an Aqualink, Pentair, etc OR it can be Nixie which is built into njsPC.  Nixie is the default unless we discover there is another controller already on the bus.  Nixie is controlling your SWG. 

0x15 isn't a Pentair pump address, however.  Pentair pumps can be set to be Pump 1-16 but when they are talking on the RS-485 bus they will be assigned 96-112 (if the system supports that many pumps).    Did you manually add a pump in the njsPC config settings?  You might also want to attach a [packet capture](https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution#-dashpanel) and we can help you see what's going on.

**Reply by zombielinux** [reply]
Alright, an update: 

After doing all sorts of packet capture and oscilloscope work. I finally broke down and called customer service. 

It turns out, these evo motors, with the single button interface, were shipped without functioning RS485. There is a beta program going around where they will replace the whole control board with a reprogrammed (and presumably, re-engineered) one. 

The rep I was in contact with seems to have ghosted me, so I'm publicly relaying the information here.

**Reply by tagyoureit** [reply]
Well, that sucks!  Thanks for keeping us posted.

**Comment by brentk7** [comment]
I have a [Zodiac FloPro VS pump](https://www.zodiac.com.au/floprovs-equipment-assistance) which uses a V-Green Motor.  This pump also includes a built in control panel which communicates with the pump via RS485. (I think it is called a Jandy FloPro VS in other parts of the world.) I was hoping to control the pump using Nixie via RS485, while still being able to use the built in control panel. I have tried configuring this as one the various VS pumps in NJSPC - but none of them work. **Would it be possible to have 2 masters controlling the pump (Nixie and the panel)?** As an alternative I am able to control the pump via relays.

I am happy to try capture the traffic between the controller and the pump if this would help get a configuration working for these pumps. **What is the best way to capture the traffic and would this be useful for anybody?**
The messages which show up in Message manager show up as "chlorinator" but it does not seem to display all the data - I f I connect the serial port to a debug node in NodeRed - I see many more packets.

**Reply by tagyoureit** [reply]
Pumps can never have two masters.  Imagine if it was told to run at 1000rpm and 2000rpm at the same time?  It just wouldn't work.  However, we can easily capture the traffic.  Which controller do you have?  

Please see [how to capture packets](https://github.com/tagyoureit/nodejs-poolController/wiki/How-to-capture-all-packets-for-issue-resolution#-dashpanel) in the wiki and do post up what you capture here.

**Reply by brentk7** [reply]
I did a capture while doing a number of operations on the pump: briefly stopping & starting at the various preset speeds (although the pump only ran at the current priming setpoint of 600rpm), and adjusting various parameters in the menus. [20230223_replay.zip](https://github.com/tagyoureit/nodejs-poolController/files/10812374/20230223_replay.zip)

It does not look like NJsPC is capturing all the packets - when I connect other capture software on a Windows PC - I see a lot more traffic than NJsPC shows. I have also attached a similar capture I made using the free trial version of [Device Monitoring Studio](https://www.hhdsoftware.com/device-monitoring-studio).
[Device Monitoring Studio Capture.zip](https://github.com/tagyoureit/nodejs-poolController/files/10812377/Device.Monitoring.Studio.Capture.zip)

**Reply by brentk7** [reply]
> Pumps can never have two masters. Imagine if it was told to run at 1000rpm and 2000rpm at the same time? It just wouldn't work. However, we can easily capture the traffic. Which controller do you have?

My pump has a built in control panel which appears to be available as a [Jandy JEP-R controller](https://www.jandy.com/en/products/pool-pumps/jep-r). 

I realise there would most likely be a conflict especially given the fact that most of these devices seem to stop if they do not receive a communication packet for s et duration. I guess I was mostly hoping to be able to "eavesdrop" on the communication to get the current state of the pump (speed, power, etc. ) to display in Home Assistant.

**Reply by rstrouse** [reply]
Jandy has more than one protocol and in the past they have been aggressive with devs looking to integrate their products.  The reason why you are seeing more data on Windows is likely because Jandy has two checksum routines depending on what type of message is encoded.  njsPC tosses these packets that it does not recognize.

**Comment by whizkidTRW** [comment]
So I just put a Century VGreen 1.65HP motor on my Hayward pump (it was an old single speed, boring motor . . .) Following the steps in comment #8 on this page at TFP: https://www.troublefreepool.com/threads/century-regal-vgreen-motor-automation.238733/, I have it speaking the Jandy protocol. I can manually control the pump speed, on, off, and get status via the appropriate hex codes. It uses address 78 by default, instead of 95 that njsPC uses. I'm going to see if I can add in support for this motor into njsPC, I have decent programming experience, but where's the best place to start?

EDIT: I'm afraid this is beyond my capabilities. I can do a lot, but probably not this. So, how can I be of help? I have this VGreen pump and a Hayward SwimPure Plus SWG that I'm ultimately trying to control with Home Assistant, via njsPC obviously, but have yet to get either working. I've also got a esp32 ATOM coming in order to try this directly from HA: https://github.com/gazoodle/CenturyVSPump.

I'd really like to help out the njsPC community and get this added since it seems the VGreen pump is a common (and easy) replacement for the OEM ones.

For reference, here are the (Jandy pump) commands I've used successfully so far:
Set speed (demand) to 2500:  10 02 78 44 00 10 00 27 05 10 03 (I got this from the TFP post, not sure how the 2500 is encoded yet)
Start:  10 02 78 41 cb 10 03
Stop:  10 02 78 42 cc 10 03
Status:  10 02 78 43 cd 10 03

**Reply by rstrouse** [reply]
If you get that pump speaking Hayward njsPC will control it without modification.

**Reply by whizkidTRW** [reply]
Unfortunately, I don't think that's possible. According to the protocol spec, it only supports the following, with the "Default" being VGreen specific using addressing like "15 43 20 50 ec" (Status).

<img width="636" alt="image" src="https://user-images.githubusercontent.com/32349482/231635779-7ef929fc-f06c-471f-a53e-08aa7aebbdfe.png">

**Reply by rstrouse** [reply]
Pentair would work as well and Prologic is Hayward.

**Reply by whizkidTRW** [reply]
Ah . . . Got it. The OP post on TFP stated that anything besides 0 (Default) or 1 (Jandy) wasn't accepted, but I will try that myself this evening to confirm and report back. Thanks for the info! I didn't know that Prologic/Omnilogic was Hayward; hadn't bothered to look those up. Thanks again!

**Reply by celestinjr** [reply]
@whizkidTRW were you able to achieve control of the v-green from njsPC?  Also, which motor model do you have?

I'm looking at the [EVQ225 (INYO link)](https://www.inyopools.com/Products/00200017088726.htm).  It [claims to be able to be controlled](https://images.inyopools.com/cloud/documents/CENTURY-CONNECT-AUTOMATION-COMPATIBILITY-CHART.pdf) from Pentair (EasyTouch, Intellitouch, Intelliconnect), Hayward (OmniLogic, ProLogic, OmniPL, EcoStar), and Jandy/Luidra (iQPump 1.0, Aqualink TCX, iAqualink 3.1, Aqualink RS).

Anyone using this motor successfully with njsPC?

**Reply by afassas** [reply]
@whizkidTRW I am hoping you can help me.  Your comment about finding the actual address of the Century VGREEN motor caught my attention, ". It uses address 78 by default, instead of 95 that njsPC uses.".

My situation is different in that I have a Jandy Aqualink TCX pool automation controller (for limited pool setups).  I successfully was able to connect my new Century EVC165 motor via RS-485 to my TCX.  I was able to control it just fine at first.  The pool motor connection caused an issue with being able to control by SWCG and I kept trying to get them to play nice together on the TCX.

I directly connected my laptop w/ USB-RS-485 dongle to my Century motor and changed the address to the MODBUS default address of 0x15.  I DON"T KNOW what the original factory address of the Century VGreen EVO motor with Century Connect was before I started.  Now I have broken the ability of the TCX to recognize or communicate with my motor.  The motor is fine otherwise and I can control it with direct MODBUS commands from the laptop.

**When you state that your motor address was found to be '78', was that the original address from the factory?**

**Reply by whizkidTRW** [reply]
Yes, I do believe that was the default. I followed Gazoodle's write up here: https://github.com/gazoodle/CenturyVSPump.

I too tried to combine the SWG RS-485 w/ the Century pump and had issues as well, so I ended up separating them. Which wasn't a problem for me, since I'm bringing it all into Home Assistant anyway. 

Good luck!

**Reply by zombielinux** [reply]
@tagyoureit 

> Pentair would work as well and Prologic is Hayward.

Neither Pentair nor Hayward seem to work or retain their settings on the motor. After following the instructions here: https://www.troublefreepool.com/threads/century-regal-vgreen-motor-automation.238733/post-2091522

The only mode that seems to latch (and thus need to be reset) is the Jandy protocol. All others drop back into the Modbus interface and don't respond to Pentair or Hayward commands.

The only command that should need to be changed is "M3" which for Pentair should be: $15$64$20$8A$05$00$02$6D$FA

and for Hayward should be: $15$64$20$8A$05$00$03$AC$3A

My reset procedure was: 
1) Turn off pump at breaker
2) Remove RS485 Connector
3) Reset dip switches to "Schedule A" (Both off)
4) Turn pump on at breaker until priming is complete (3 minutes)
5) Turn pump off at breaker
6) Re-attach RS485 connector
7) Reset dip switches to "Schedule D" (Both on)

I'm willing to help get the Jandy or Modbus protocols supported on this motor, as it seems there's quite a number of people looking for help and zero manufacturer support.

**Comment by whizkidTRW** [comment]
No, not from njsPC. I have the [Century V-Green, Square Flange, 48Y motor](https://www.inyopools.com/Products/00200017074679.htm?gad_source=1&gclid=CjwKCAjw3NyxBhBmEiwAyofDYfJN28RbomX8RaNxAy52eR-I1ex3k1YXTpkHpBIqgYkQYyhL3_P7txoC-DIQAvD_BwE) which was a perfect fit for my Hayward pump. I ended up following [Gazoodle's code](https://github.com/gazoodle/CenturyVSPump) and controlling it with his suggested M5Stack ESP32 WiFi RS485 chip. It works great!

In the end, since I'm using Home Assistant as the controller, njsPC is really just an interface to the Hayward SwimPure plus SWG. What I eventually figured out was that the previous owners had replaced the SWG logic board at some point with a knockoff/generic Chinese version that DOES NOT have RS485 control onboard. I found a fried genuine Hayward board that I was able to repair and it's been working for the past year.

The motor you linked appears to run the same Century control protocol as mine, so I'd think it would work. But, again, I'm using Home Assistant to bring everything together since Gazoodle's code feeds into HA via ESPHome, so both systems are able to be controller from one dashboard in HA. Here's my HA dashboard for "outside", for what it's worth. Best of luck!

![image](https://github.com/tagyoureit/nodejs-poolController/assets/32349482/1e95cb83-3aa5-4c11-a078-8f840fb197d9)

**Comment by mguinness** [comment]
Support for Century modbus was added in https://github.com/tagyoureit/nodejs-poolController/pull/1120
